### PR TITLE
bdDetect: Enrich USB serial data with info about USB device

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Blocks #15693

### Summary of the issue:
Many braille displays use a serial converter from FTDI with VID 0x403" and PID 0x6001. Given the weird nature of Albatross devices not being able to identify themselves on request, the Albatross driver is currently a disturbing factor in automatic detection.

### Description of user facing changes
None.

### Description of development approach
1. This pr enriches information for serial ports in bdDetect. The _DeviceInfoFetcher class now has an extra autoprop usbComPorts that bundles com port info with USB metadata. If all works as expected, the Albatross driver will be able to limit itself to Albatross devices only after this pr by using the `busReportedDeviceDescription` dictionary entry.
2. Removed duplicated code from hwPortUtils

### Testing strategy:
Testing with Albatross from @burmancomp would be appreciated.
It can be tested using `bdDetect.getConnectedUsbDevicesForDriver("albatross")`. This should yield device matches with details about the underlying USB device.

### Known issues with pull request:
None known

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by burmancomp** [comment]
Yes, I am about to test this likely tomorrow.

**Comment by burmancomp** [comment]
Portable version of nvda_snapshot_pr15764-29937,d9a7c5be.exe, python console:

>>> next(bdDetect.getConnectedUsbDevicesForDriver("albatross"))
DeviceMatch(type='serial', id='VID_0403&PID_6001', port='COM5', deviceInfo={'hardwareID': 'USB\\VID_0403&PID_6001&REV_0400', 'port': 'COM5', 'usbID': 'VID_0403&PID_6001', 'friendlyName': 'USB Serial Port (COM5)', 'devicePath': '\\\\?\\usb#vid_0403&pid_6001#tmboijcd#{a5dcbf10-6530-11d2-901f-00c04fb951ed}', 'deviceDescription': 'USB Serial Converter'})
Let me know did I test correctly?

**Comment by leonardder** [comment]
Yes, this is correct. I need to expand hwPortUtils but I at least know how.
Could you by any chance do the following?

1. Open Device Manager
2. Open `ports (com & lpt)`
3. Select your com5 port
4. Look up de device details and in there, the Bus reported device description
I assume that's USB Serial Port, right?

**Comment by burmancomp** [comment]
Yes, it is "USB Serial Port", and usb serial converter bus reported device description is "Albatross Braille Display".

Let me know if you need any additional information.

**Comment by leonardder** [comment]
I think the most recent commit should finally give us what's needed, namely `Albatross Braille Display` in the `busReportedDeviceDescription`

**Comment by burmancomp** [comment]
Unfortunately albatross did not start, and it is not present in list of displays.

>>> import bdDetect
>>> next(bdDetect.getConnectedUsbDevicesForDriver("albatross"))
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "bdDetect.pyc", line 441, in getConnectedUsbDevicesForDriver
  File "baseObject.pyc", line 62, in __get__
  File "baseObject.pyc", line 168, in _getPropertyViaCache
  File "bdDetect.pyc", line 245, in _get_usbDevices
  File "hwPortUtils.pyc", line 424, in listUsbDevices
OSError: [WinError 1168] Elementtiä ei löydy.

**Comment by leonardder** [comment]
Have you performed the test step on the python console?

**Comment by burmancomp** [comment]
Yes I tested in python console getting those result.

**Comment by leonardder** [comment]
Ah I'm sorry, somehow I didn't get your test result.

**Comment by leonardder** [comment]
Pretty sure last commit eventually should get you going

**Comment by burmancomp** [comment]
congratulations @leonardder!

`>>> import bdDetect
>>> next(bdDetect.getConnectedUsbDevicesForDriver("albatross"))
DeviceMatch(type='serial', id='VID_0403&PID_6001', port='COM5', deviceInfo={'hardwareID': 'USB\\VID_0403&PID_6001&REV_0400', 'port': 'COM5', 'usbID': 'VID_0403&PID_6001', 'friendlyName': 'USB Serial Port (COM5)', 'devicePath': '\\\\?\\usb#vid_0403&pid_6001#tmboijcd#{a5dcbf10-6530-11d2-901f-00c04fb951ed}', 'busReportedDeviceDescription': 'Albatross Braille Display'})
`

I am about to close #15693.

**Comment by leonardder** [comment]
Note that this pr doesn't yet contain a workaround for Albatross. Therefore I'd leave #15693 open and adapt it after merge of this pr.

**Comment by burmancomp** [comment]
Thank you. I leave it open.

**Comment by burmancomp** [comment]
> Thank you for information. I leave it open.

ps. After reading your previous comment and then inspecting your initial comment I realized that your thought is that albatross driver is changed too. I realized it before writing my previous comment (which I am quoting in this comment).

**Comment by leonardder** [comment]
This pr does not yet improve the Albatross driver by itself.

**Comment by burmancomp** [comment]
But albatross driver can query bus reported device description?

**Comment by leonardder** [comment]
Sure, after this is merged.

File: source/hwPortUtils.py [hunk 1]
```diff
@@ -127,147 +131,115 @@ def _isDebug():
 
 
 # C901 'listComPorts' is too complex. Look for opportunities to break this method down.
-def listComPorts(onlyAvailable=True) -> typing.Iterator[typing.Dict]:  # noqa: C901
+def listComPorts(onlyAvailable: bool = True) -> typing.Iterator[dict]:  # noqa: C901
 	"""List com ports on the system.
```
Comment (seanbudd) [inline source/hwPortUtils.py:135]: considering this is a full rewrite, the complexity should be reduced to remove the c901. I'd encourage splitting this up into helper functions

File: source/hwPortUtils.py [hunk 2]
```diff
@@ -127,147 +131,115 @@ def _isDebug():
...
+			log.debugWarning(f"Couldn't get SPDRP_FRIENDLYNAME for {port!r}: {ctypes.WinError()}")
+			entry["friendlyName"] = port
+		else:
+			entry["friendlyName"] = buf.value
 
-			if _isDebug():
-				log.debug("%r" % entry)
-			yield entry
+		if _isDebug():
+			log.debug("%r" % entry)
```
Comment (seanbudd) [inline source/hwPortUtils.py:234]: could this be updated to use f-string

File: source/hwPortUtils.py [hunk 3]
```diff
@@ -337,18 +309,20 @@ def getWidcommBluetoothPortInfo(port):
...
 
 
-def listUsbDevices(onlyAvailable=True) -> typing.Iterator[typing.Dict]:
-	"""List USB devices on the system.
-	@param onlyAvailable: Only return devices that are currently available.
-	@type onlyAvailable: bool
-	@return: Generates dicts including keys of usbID (VID and PID), devicePath and hardwareID.
+def _listDevices(
+		deviceClass: GUID,
+		onlyAvailable=True
```
Comment (seanbudd) [inline source/hwPortUtils.py:314]:
```suggestion
		onlyAvailable: bool = True
```

File: source/hwPortUtils.py [hunk 4]
```diff
@@ -396,33 +371,63 @@ def __str__(self):
...
-				if _isDebug():
-					log.debug("%r" % usbId)
-				yield info
+			yield (g_hdi, idd, devinfo, buf)
+
 	finally:
 		SetupDiDestroyDeviceInfoList(g_hdi)
+
+
+def listUsbDevices(onlyAvailable=True) -> typing.Iterator[dict]:
```
Comment (seanbudd) [inline source/hwPortUtils.py:380]:
```suggestion
def listUsbDevices(onlyAvailable: bool = True) -> typing.Iterator[dict]:
```

File: source/hwPortUtils.py [hunk 5]
```diff
@@ -396,33 +371,63 @@ def __str__(self):
...
+		else:
+			# The string is of the form "usb\VID_xxxx&PID_xxxx&..."
+			usbId = buf.value[4:21]  # VID_xxxx&PID_xxxx
+			entry.update({
+				"hardwareID": buf.value,
+				"usbID": usbId,
+				"devicePath": idd.DevicePath
+			})
+			if _isDebug():
+				log.debug("%r" % usbId)
```
Comment (seanbudd) [inline source/hwPortUtils.py:408]: can this use an f-string

File: source/hwPortUtils.py [hunk 6]
```diff
@@ -490,7 +495,7 @@ def _getHidInfo(hwId, path):
 _hidGuid = None
 
 
-def listHidDevices(onlyAvailable=True) -> typing.Iterator[typing.Dict]:
+def listHidDevices(onlyAvailable=True) -> typing.Iterator[dict]:
 	"""List HID devices on the system.
 	@param onlyAvailable: Only return devices that are currently available.
 	@type onlyAvailable: bool
```
Comment (seanbudd) [inline source/hwPortUtils.py:501]:
```suggestion
```

File: source/hwPortUtils.py [hunk 7]
```diff
@@ -490,7 +495,7 @@ def _getHidInfo(hwId, path):
 _hidGuid = None
 
 
-def listHidDevices(onlyAvailable=True) -> typing.Iterator[typing.Dict]:
+def listHidDevices(onlyAvailable=True) -> typing.Iterator[dict]:
```
Comment (seanbudd) [inline source/hwPortUtils.py:498]:
```suggestion
def listHidDevices(onlyAvailable: bool = True) -> typing.Iterator[dict]:
```

File: user_docs/en/changes.t2t [hunk 8]
```diff
@@ -127,6 +129,14 @@ That method receives a ``DriverRegistrar`` object on which the ``addUsbDevices``
 - ``languageHandler.makeNpgettext`` and ``languageHandler.makePgettext`` have been removed.
 ``npgettext`` and ``pgettext`` are supported natively now. (#15546)
 - The app module for [Poedit https://poedit.net] has been changed significantly. The ``fetchObject`` function has been removed. (#15313, #7303, @LeonarddeR)
+- The following redundant types and constants have been removed from ``hwPortUtils``:  (#15764, @LeonarddeR)
+ - ``PCWSTR``
+ - ``HWND`` (Replaced by ``ctypes.wintypes.HWND``)
+ - ``ULONG_PTR``
+ - ``ULONGLONG``
+ - ``NULL``
+ - ``GUID`` (Replaced by ``comtypes.GUID``)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:138]:
```suggestion
 - ``HWND`` (replaced by ``ctypes.wintypes.HWND``)
 - ``ULONG_PTR``
 - ``ULONGLONG``
 - ``NULL``
 - ``GUID`` (replaced by ``comtypes.GUID``)
```

File: user_docs/en/changes.t2t [hunk 9]
```diff
@@ -127,6 +129,14 @@ That method receives a ``DriverRegistrar`` object on which the ``addUsbDevices``
 - ``languageHandler.makeNpgettext`` and ``languageHandler.makePgettext`` have been removed.
 ``npgettext`` and ``pgettext`` are supported natively now. (#15546)
 - The app module for [Poedit https://poedit.net] has been changed significantly. The ``fetchObject`` function has been removed. (#15313, #7303, @LeonarddeR)
+- The following redundant types and constants have been removed from ``hwPortUtils``:  (#15764, @LeonarddeR)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:132]: just confirming, have you searched NVDA for usage of these? there doesn't appear to be any wider changes in this diff.

Comment (leonardder) [reply user_docs/en/changes.t2t:132]: Yes I have. These constants were probably copied from the scanner in the pyserial module.

**Comment by leonardder** [comment]
The only way I could get rid of the c901 was by rewriting the if/elif statements into a match statement.

**Comment by lukaszgo1** [comment]
> The only way I could get rid of the c901 was by rewriting the if/elif statements into a match statement.

That works only because `match` and `case` statements are not measured for complexity by mccabe, so the function is as complex as before, the only change is that complexity is not checked for a language construct used. If it is indeed so hard to simplify it perhaps the `noqa` comment should remain, so that it is clear we haven't really gained anything here?

**Comment by leonardder** [comment]
In that case I prefer to revert the match statement.

**Comment by seanbudd** [comment]
>so the function is as complex as before

I don't think this is true - the function was still split up into listComPorts/_getBluetoothPortInfo which reduced complexity in listComPorts significantly, and I am fairly sure to the point of legitimate c901 removal in both functions.

**Review by seanbudd — APPROVED** [review]
Thanks @leonardder

**Pull request merged**